src/main/java/com/maxdemarzi/shortest/Dijkstra.java

```diff
@@ -0,0 +1,201 @@
+
+        public Step(long nodeId, int cost) {
+            this.nodeId = nodeId;
+            this.cost = cost;
+        }
+
+        public final int compareTo(Step o) {
+            if (this.cost < o.cost) {
+                return -1;
+            } else if (this.cost > o.cost) {
+                return 1;
+            }
+            return 0;
+        }
+
+        public boolean isStartNode() {
+            return false;
+        }
+    }
+
+    public Dijkstra(ReadOperations readOps, IntIntMap relationshipCosts, Map<Long, Integer> startNodes, int maxCost, NodeCallback callback) {
+        super();
+        this.readOps = readOps;
+        this.relationshipCosts = relationshipCosts;
+        this.nodeCallback = callback;
+        this.queue = new PriorityQueue<Step>();
+        this.paths = HashLongLongMaps.newMutableMap(500);
+        this.maxCost = maxCost;
+        //initialize the queue with our start nodes
+        for (Map.Entry<Long,Integer> entry : startNodes.entrySet()) {
+            long nodeId = entry.getKey().longValue();
+            int cost = entry.getValue().intValue();
+
+            Step step = new Step(nodeId, cost) {
+                public boolean isStartNode() {
+                    return true;
+                }
+            };
+            this.paths.put(nodeId, costPaths(cost, 1));
+            this.queue.offer(step);
+        }
+    }
+
+    public void step() {
+        final Step current = this.queue.poll();
+        if (current == null) {
+            this.finish();
+            return;
+        }
+        final long exploredCostPaths = this.paths.get(current.nodeId);
+        // Because we cant modify the priority of things on the queue, we just have to handle
+        // encountering duplicates. removing and re-inserting would be super slow
+        if (explored(exploredCostPaths)) {
+            return;
+        }
+
+        final int paths = paths(exploredCostPaths);
+        final int cost = cost(exploredCostPaths);
+
+        final Cursor<NodeItem> nodeCursor = this.readOps.nodeCursor(current.nodeId);
+        nodeCursor.next();
+        final NodeItem currentNode = nodeCursor.get();
+        final Cursor<RelationshipItem> relationshipCursor = currentNode.relationships(Direction.BOTH);
+        int degree = 0;
+
+        if (cost != this.maxCost) { // if we're at max cost, dont bother looking at edges
+            while(relationshipCursor.next()) {
+                degree++;
+                RelationshipItem relation = relationshipCursor.get();
+                final int stepCost = this.relationshipCosts.get(relation.type()) + cost;
+                if (stepCost > this.maxCost) {
+                    continue;
+                }
+                final long otherId = relation.otherNode(current.nodeId);
+                final long newVal = costPaths(stepCost, paths);
+
+                final long result = this.paths.merge(otherId, newVal, updateSeenFunc);
+                if (result == newVal) {
```

Comment:
if we added paths to something with the old cost, we don't put it back on the queue - is that right?